### ISSUE TITLE
feature request: reporting

### ISSUE THREAD

**Issue by johanpeters** [akeeba/panopticon#220]
something for the roadmap would be nice to have reporting based on for instance montly bases and a report for avery website (setting enable / disable report).
inside report the statists of the last month regarding updates, success, failure, status, etc.

**Comment by nikosdion** [comment]
This does not sound like it serves any real purpose. If there is a failure you need to address it when it happens, not a month later. 

Moreover, collecting this information takes up a substantial amount of space and eventually slows the database down to a crawl. To address that, we'd have to either only store a rolling log (keep the last month until we email you about it, then remove the old data), or implement a very complex scheme with two databases (hot / cold data) which means that Panopticon is incompatible with a large number of server configurations.

What exactly is your use case here?

**Comment by brianteeman** [comment]
I would guess that a use case for this would be to send a report each month to a client on the work that has been done on their site as evidence that maintenance is taking place

**Comment by nikosdion** [comment]
@brianteeman And you need to log failures, too?

**Comment by brianteeman** [comment]
Not in that use case.

**Comment by nikosdion** [comment]
I am thinking that it would be best if the update action logging toggle had a default global value and a per site value. It would actually be two toggles:

- Log CMS updates for reporting: None, Successful, Unsuccessful, Both
- Log extension updates for reporting: None, Successful, Unsuccessful, Both

The default setting will be None, as it is a feature which can crash the database in some cases. I will probably also need to set up a scheduled task for exporting the fundamental tables' contents (users, groups, sites) in case the database goes ka-boom and you need to restore it on the double-fast.

I believe that in the @brianteeman use case you'd want to use the Successful setting either globally or for each site, and for @johanpeters use case you'd want to use the Both setting globally.

You could then view this report for all sites, or just specific sites, for any length of time you want with the default being the previous month.

I _probably_ can also make a print style to format this a tad prettier so you can print to PDF.

Caveat: since the extensions installed are not a constant over time I'd have to also log the extension name which is the reason this report table will get big. So I'm going to give you the option to also delete entries and probably set up a weekly task to do MySQL table maintenance (basically: a “defrag”) to ensure the table will be far less likely to crash on you.

What do you guys think of that approach?

**Comment by johanpeters** [comment]
@nikosdion : regarding report in PDF or format to print in PDF i totally agree. and basicly my use case would be the same as @brianteeman . mainly to report website owners of the things that has been done the last period.
i understand that keeping the information in de database is not effective.. so after report / schedule to cleanup the database.
lets see, report on montly base, so cleen up can be done for everything before last month.
for me it would be enough to have only reports saved in panopticon.. i don't know if it is possible to cleanup the records that are reported.. so that you never lose information thats not reported (with addition option to clear database per website based on given date

**Comment by brianteeman** [comment]
Just to be clear - I don't have a need for this myself - I was just trying to explain what you meant as I have seen it before on other platforms

**Comment by nikosdion** [comment]
Would it also make sense to detect installed and uninstalled extensions and log them into the report table as well?

**Comment by brianteeman** [comment]
Definitely log changes in the time period

**Comment by nikosdion** [comment]
Well, they would be logged as they are discovered which, by default, is once an hour. If you uninstall and reinstall an extension within that hour between consecutive extensions reload by Panopticon you wouldn't get an entry. 

That's the only reasonable thing I can do without obliterating the performance of your sites (the other solutions are checking more often for updates, of have the site itself call an API endpoint in Panopticon to report extension installation / uninstallation; both would slow down the site considerably).

**Comment by morphinestyle** [comment]
I really like this feature.

It would be fantastic to have a REPORTS section where you can insert one (or more) customer emails so that they can receive reports on what has changed on the site during the month.

**Comment by nikosdion** [comment]
Let's try to make this a user accessible report first, then we can automate it with emails and such :)

**Comment by johanpeters** [comment]
> Well, they would be logged as they are discovered which, by default, is once an hour. If you uninstall and reinstall an extension within that hour between consecutive extensions reload by Panopticon you wouldn't get an entry.
> 
> That's the only reasonable thing I can do without obliterating the performance of your sites (the other solutions are checking more often for updates, of have the site itself call an API endpoint in Panopticon to report extension installation / uninstallation; both would slow down the site considerably).

i think report of changes based on detection once every hour would be more than enough.
great thinking allong to expand the sollution.

**Comment by johanpeters** [comment]
> Let's try to make this a user accessible report first, then we can automate it with emails and such :)

agree.. first things first.. and in de mean time i am busy with translating into dutch..

**Comment by nikosdion** [comment]
I will be tracking this in #245 and I am therefore converting this issue into a discussion.
